lib/assertions.js

```diff
@@ -811,19 +811,38 @@ module.exports = function (expect) {
         return this.shift(expect, subject.apply(subject, args), 1);
     });
 
-    expect.addAssertion('array', 'when passed as parameters to [async]', function (expect, subject, fn) { // ...
+    function instantiate(Constructor, args) {
+        function ProxyConstructor() {
+            return Constructor.apply(this, args);
+        }
+        ProxyConstructor.prototype = Constructor.prototype;
+        return new ProxyConstructor();
+    }
+
+    expect.addAssertion('array', 'when passed as parameters to [async] [constructor]', function (expect, subject, fn) { // ...
         this.errorMode = 'nested';
-        if (this.flags.async) {
-            var that = this;
-            return expect.promise(function (run) {
-                var args = subject.concat(run(function (err, result) {
+        var that = this,
+            args = subject;
+        return expect.promise(function (run) {
+            if (that.flags.async) {
+                args = [].concat(args);
+                args.push(run(function (err, result) {
                     expect(err, 'to be falsy');
                     return that.shift(expect, result, 1);
                 }));
-                fn.apply(null, args);
-            });
-        } else {
-            return this.shift(expect, fn.apply(fn, subject), 1);
-        }
+                if (that.flags.constructor) {
+                    instantiate(fn, args);
+                } else {
```

Comment:
I know it's a bit of a weird case, but I think I have seen such creatures. But to be honest I mostly implemented it so that all combinations of the two flags would do something non-surprising.